nano/node/transport/tcp.cpp

```diff
@@ -475,25 +475,52 @@ std::unique_ptr<nano::container_info_component> nano::transport::tcp_channels::c
 	return composite;
 }
 
-void nano::transport::tcp_channels::purge (std::chrono::steady_clock::time_point const & cutoff_a)
+void nano::transport::tcp_channels::purge (std::chrono::steady_clock::time_point cutoff_deadline)
 {
 	nano::lock_guard<nano::mutex> lock{ mutex };
 
-	// Remove channels with dead underlying sockets
-	erase_if (channels, [] (auto const & entry) {
-		return !entry.channel->alive ();
-	});
+	node.logger.debug (nano::log::type::tcp_channels, "Performing periodic channel cleanup, cutoff: {}s", nano::log::seconds_delta (cutoff_deadline));
+
+	auto should_close = [this, cutoff_deadline] (auto const & channel) {
+		// Remove channels that haven't successfully sent a message within the cutoff time
+		if (auto last = channel->get_last_packet_sent (); last < cutoff_deadline)
+		{
+			node.logger.debug (nano::log::type::tcp_channels, "Closing idle channel: {} (idle for {}s)",
+			channel->to_string (),
+			nano::log::seconds_delta (last));
+
+			return true; // Close
+		}
+		// Check if any tcp channels belonging to old protocol versions which may still be alive due to async operations
+		if (channel->get_network_version () < node.network_params.network.protocol_version_min)
+		{
+			node.logger.debug (nano::log::type::tcp_channels, "Closing channel with old protocol version: {}", channel->to_string ());
+
+			return true; // Close
+		}
+		return false;
+	};
 
-	auto disconnect_cutoff (channels.get<last_packet_sent_tag> ().lower_bound (cutoff_a));
-	channels.get<last_packet_sent_tag> ().erase (channels.get<last_packet_sent_tag> ().begin (), disconnect_cutoff);
+	for (auto const & entry : channels)
+	{
+		if (should_close (entry.channel))
+		{
+			entry.channel->close ();
+		}
+	}
+
+	erase_if (channels, [this] (auto const & entry) {
```

Comment:
Can these two loops be merged to close and erase simultaneously?

Reply:
No, socket closure is executed on a strand so there is no guarantee that a closed channel will be immediately dead.